pkg/resource/deploy/plan_executor.go

```diff
@@ -168,6 +168,7 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
 
 	// Set up a step generator for this plan.
 	pe.stepGen = newStepGenerator(pe.plan, opts, updateTargetsOpt, replaceTargetsOpt)
+	pe.plan.current = pe.stepGen.resourceStates
```

Comment:
a couple of things:
1. if the stepGen already gives access to the resourceStates, and the executor gives access to the stepGen, why do we need to place this on the plan itself?
2. i presume this means that resourceStates is mutable and changes over time.  Is htat ok?  are there any concerns about the semantics of that as we're running the user program?  i'm just always nervous about accessing shared mutable state

Reply:
> if the stepGen already gives access to the resourceStates, and the executor gives access to the stepGen, why do we need to place this on the plan itself?

The PlanExecutor is not available to the built-ins.  Only the Plan.  So there is no link between these.  Doing this was signfiicantly cleaner than trying to give the built-ins access to the executor.  If @pgavlin thinks that would be better - happy to do that though.

> i presume this means that resourceStates is mutable and changes over time. Is htat ok? are there any concerns about the semantics of that as we're running the user program? i'm just always nervous about accessing shared mutable state

Yes - it is mutable (and fundamentally needs to be mutable as we want to look up the current registration of a URN at a point in time).  There is certainly a contract here that our attempt to look it up happens "after" the resource is registered, and we are relying on the idea that awaiting all the channel/rpc/promise steps on the registration will ensure that the place we do the lookup is indeed "after".  It is certainly possible that there is some potential issue here - but I don't know exactly what it would be.  This is related to the question of whether the builtin functions should maintain there own state of the registered resources (though that seems like it still would have the same fundamental concern, and just more duplicated state).